### PR TITLE
Testsuite: Speed up CoreUGens tests

### PR DESCRIPTION
**Pull request by mtmccrea**
## Purpose and Motivation

In many unit tests, wait times are ad hoc, which accumulates quite a bit of time when lots of tests are run. This also points to uncertainties about how the tests are scheduled and run.

This PR tries to make the logic more strict and rational, and effectively reduces the run time from ~115 sec to ~45 sec. The “test on x” CI jobs take between 13-15 minutes for each of Mac and Linux tests. So if this approach is robust, it may be able to be applied to tests more broadly.

Improvements so far:
- add a `classvar` to have a default number of blocks to run for tests. Defaults to 10 blocks. (It’s still up to the test author to use this variable.)
- The largest improvement comes from revising the `wait` scheme
    - Removed random wait times.
    - Rely on `server.sync` to throttle looping during asynchronous testing
        - In these cases, the approach of counting the `completed` tests for the predicate of the waiting condition, and in those cases an additional check is made to ensure all tests completed.
        - An edge case appears to be when only one (or some small number of tests?) is run in a loop, the wait condition times out. The timeout time is calculated based on `testDuration x numTests x 1.5`, so when only one test runs with a very short duration (say, 10 blocks, the default) it times out because the time for resource allocation is comparable to the test run time. So, there is a minimum timeout of 2 seconds imposed. This is just a failsafe, as none of these test loops should ever time out, and an error is thrown if it does.
- `Pitch` and `ZeroCrossing` tests were rewritten to be easier to reason without magic numbers, and in turn now pass with higher accuracy and shorter run times. (were 10 sec each)
- `Out.ar` was removed from these tests so there is no sound out during tests
        - For tests that do use output (e.g. `RecordBuf` uses `.play` to write to a buffer), the `outbus` is set to `server.options.numInputBusChannels + server.options.numOutputBusChannels`, i.e. the first virtual audio bus, so no sound is written to hardware.
- `report` was set to `false` throughout. This would otherwise create trivial noise in the output of passing tests. Perhaps someone can provide some rationale for setting this to `true`.

## Types of changes

- Tests

## To-do list

- [x] Code is tested
- [ ] All tests are passing
- [ ] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

File: testsuite/classlibrary/TestCoreUGens.sc [hunk 1]
```diff
@@ -309,125 +316,143 @@ TestCoreUGens : UnitTest {
...
-				completed = completed + 1;
-				condvar.signalOne;
-			});
-			0.32.wait;
+				RecordBuf.ar(PlayBuf.ar(numchans, dataBuf, BufRateScale.ir(dataBuf), doneAction: 2), recBuf, loop:0) * 0.1;
+			}.play(server, outbus: server.options.numInputBusChannels + server.options.numOutputBusChannels); // avoid hw output: write to first virutal bus
+
 			server.sync;
+
+			fork {
```
Comment (mtmccrea) [inline testsuite/classlibrary/TestCoreUGens.sc:424]: Would be good to have feedback on this: a routine is embedded within the test. I _think_ this is ok because the timing condition is waiting for the `completed` variable update, which is shared between the main routine and this forked routine.

Comment (telephon) [reply testsuite/classlibrary/TestCoreUGens.sc:424]: yes, it should be ok. A `forkIfNeeded` would embed the test if it happens in  a routine already.

Comment (telephon) [reply testsuite/classlibrary/TestCoreUGens.sc:424]: actually, I was wrong, better *not* use a forkIfNeeded.

Comment (telephon) [reply testsuite/classlibrary/TestCoreUGens.sc:424]:
```

// works as expected
(
fork {

	var con = CondVar.new;
	fork {
		2.wait;
		con.signalOne;
	};
	con.wait;
	"hello".postln
	
}
)


// doesn't work correctly
(
fork {

	var con = CondVar.new;
	forkIfNeeded {
		2.wait;
		con.signalOne;
	};
	con.wait;
	"hello".postln
	
}
)
```

File: testsuite/classlibrary/TestCoreUGens.sc [hunk 2]
```diff
@@ -309,125 +316,143 @@ TestCoreUGens : UnitTest {
...
+				(bufDur * 1.05).wait; // wait for the buffer to fill up
+				recBuf.loadToFloatArray(action: { |renderedData|
```
Comment (telephon) [inline testsuite/classlibrary/TestCoreUGens.sc:426]: Now there is a method `asBuffer` in function. This may simplify the logic a little bit here.

File: testsuite/classlibrary/TestCoreUGens.sc (hunk 1 above)
Comment (mtmccrea) [reply testsuite/classlibrary/TestCoreUGens.sc:424]: Yes, you can think of the routine being "flattened" to remove the inner fork:
```supercollider
// doesn't work correctly
(
fork {

	var con = CondVar.new;
	forkIfNeeded {
		"waiting".postln;
		2.wait;
		"waited, signalling".postln;
		con.signalOne;
	};
	"waiting outside".postln;
	con.wait;
	"hello".postln
}
)
```
So it signals the condition before the thread is made to wait on the condition.

Comment (telephon) [reply testsuite/classlibrary/TestCoreUGens.sc:424]: yes, we could mention this in `forkIfNeeded` sometime.

**Review by telephon — APPROVED** [review]

File: testsuite/classlibrary/TestCoreUGens.sc (hunk 2 above)
Comment (mtmccrea) [reply testsuite/classlibrary/TestCoreUGens.sc:428]: In this case we're testing the data written to a preallocated buffer via `RecordBuf`,  as opposed to testing the output of the "synth" (function). Unless you're referring to a different function?

**Comment by mtmccrea** [comment]
Just did some minor cleanup of comments and variables that were replaced.

File: testsuite/classlibrary/TestCoreUGens.sc (hunk 2 above)
Comment (telephon) [reply testsuite/classlibrary/TestCoreUGens.sc:428]: yes, just in that case, of the output of a synth.

**Comment by telephon** [comment]
Just a guess: maybe this needs to be higher for the TestUGen_RTAlloc ugens?

```
server.options.memSize = 1024;
```

**Comment by dyfer** [comment]
> Just a guess: maybe this needs to be higher for the TestUGen_RTAlloc ugens?

Hm, that might be it: https://github.com/supercollider/supercollider/actions/runs/9902226869/job/27358145553?pr=6374#step:6:14560

**Comment by telephon** [comment]
Exactly, that's what I mean.

**Comment by mtmccrea** [comment]
> > Just a guess: maybe this needs to be higher for the TestUGen_RTAlloc ugens?
> 
> Hm, that might be it: [`LocalBuf_allocBuffer: alloc failed, increase server's RT memory (e.g. via ServerOptions)`](https://github.com/supercollider/supercollider/actions/runs/9902226869/job/27358145553?pr=6374#step:6:14560)

https://scsynth.org/chat/c/-/45/3280

**Comment by capital-G** [comment]
@mtmccrea can you rebase this against develop so we can see if we get a successful CI? RTAlloc tests are now temporarily disabled in develop b/c they create problems.

**Review by capital-G — APPROVED** [review]
Thanks for rebasing! The PR is really appreciated, anything that makes the tests more deterministic and stable is a great help!

The changes look good from my side, as @telephon has some unresolved comments in this PR, he should merge it.

File: testsuite/classlibrary/TestCoreUGens.sc (hunk 2 above)
Comment (mtmccrea) [reply testsuite/classlibrary/TestCoreUGens.sc:428]: @telephon, to clarify whether this comment is resolved, I don't see that `asBuffer` is useful in this case, since there is no synth output being generated. The synth under test only uses `RecordBuf`, not `Out` and it's the contents of the buffer being written to that is tested against `data`.

I see how 
```
// random data for test
data = { 1.0.rand }.dup((server.sampleRate * bufDur).round * numchans);
// load data to server
dataBuf = Buffer.loadCollection(server, data, numchans);
```
could make use of `asBuffer` to be more concise, but the `data` array is used later
```
this.assertArrayFloatEquals(renderedData - data, ///
```
so if a buffer were loaded directly with `asBuffer` it would just need to be unpacked later for the assert test.

**Comment by mtmccrea** [comment]
Rebased again after conflict with #6441, which was actually redundant with this PR.

**Comment by mtmccrea** [comment]
No problem, thanks for reviewing!

**Comment by capital-G** [comment]
I'll merge it - hopefully it improves the state of the tests that #6441 caused 😶‍🌫️

File: testsuite/classlibrary/TestCoreUGens.sc [hunk 3]
```diff
@@ -309,124 +318,143 @@ TestCoreUGens : UnitTest {
...
+
 		tests.keysValuesDo{|name, func|
-			func.loadToFloatArray(0.01, server, { |data|
+			func.loadToFloatArray(testDur, server, { |data|
 				this.assertArrayFloatEquals(data, 0, name.quote, report: true);
 				completed = completed + 1;
-				condvar.signalOne;
+				cond.signalOne;
 			});
+			server.sync;
```
Comment (JordanHendersonMusic) [inline testsuite/classlibrary/TestCoreUGens.sc:391]: @mtmccrea I was just wondering (after #6441) if  you could explain why this `server.sync` is needed? I can't seem to why this passes, but the other approach fails.

Comment (JordanHendersonMusic) [reply testsuite/classlibrary/TestCoreUGens.sc:391]: Ah I think I've figured this out. If you switch to TCP and remove the syncs, you don't need this and performance increases by a factor of 2. I suppose this means the udp messages are getting dropped.

I've also found you get a significant performance increase (circa 5x) in a tight loop by only syncing every 6 iterations (UDP)
```
if (n % 6 == 0) { s.sync };
```

What do you think about switching all these tests over to use TCP and removing the s.sync? It passes on my machine.

---

Sorry for not commenting on this when it was still open!

File: testsuite/classlibrary/TestCoreUGens.sc [hunk 4]
```diff
@@ -309,124 +318,143 @@ TestCoreUGens : UnitTest {
...
 		};
 
-		condvar.waitFor(1, { completed == tests.size });
+		testsFinished = cond.waitFor((1.5 * tests.size * testDur).max(minTimeOut), { completed == tests.size });
```
Comment (JordanHendersonMusic) [inline testsuite/classlibrary/TestCoreUGens.sc:394]: I just had a question about your intent here. The time here is the time between each `condvar.signalOne`, not the total time of waiting (documentation seems lacking here), is it correct to multiply by the number of tests in this case  — I mean it works so it doesn't really matter.

File: testsuite/classlibrary/TestCoreUGens.sc (hunk 3 above)
Comment (mtmccrea) [reply testsuite/classlibrary/TestCoreUGens.sc:391]: Seems like a great thing to propose in a new PR, so others can weigh in if they know more useful info about UDP vs. TCP behavior. I'm surprised UDP gets dropped on an internal messaging. I'm not sure where the point of failure is vs. TCP... does TCP  just have a larger message queue?

One of the goals of this PR was to model a fairly principled approach to timing tests, which might then find its way into more areas of the test suite. So if we can pin down in some degree of detail why TCP allows higher throughput, we could document it as best practices.

Looks promising!

File: testsuite/classlibrary/TestCoreUGens.sc (hunk 4 above)
Comment (mtmccrea) [reply testsuite/classlibrary/TestCoreUGens.sc:394]: Thanks for looking at this in detail. The way to think about this is the timeout is set to a duration beyond the completion time of _all_ tests. It's a bit confusing because this condition is signaled after each test, but it's actually waiting for the predicate `completed == tests.size` to be `true`. From the [docs](https://docs.supercollider.online/Classes/CondVar.html#-waitFor):

>If predicate is not `nil`, this method returns immediately if `predicate.value` is `true`, and otherwise blocks either until the timeout expires, or until the thread is woken by `signalOne` or `signalAll` and `predicate.value` is `true`.

The mechanics of `CondVar` take some getting used to (I have to re-read docs every time I use it).

Comment (JordanHendersonMusic) [reply testsuite/classlibrary/TestCoreUGens.sc:394]: Thanks for this! I actually wrote some tests before I wrote that to try and make sure what I was saying was correct... I guess I messed up somewhere as I can't repeat it now! Thanks again!

File: testsuite/classlibrary/TestCoreUGens.sc (hunk 3 above)
Comment (mtmccrea) [reply testsuite/classlibrary/TestCoreUGens.sc:391]: @dyfer linked to @Spacechild1's helpful post [here](https://github.com/supercollider/supercollider/issues/5870#issuecomment-1310965786).

So, if we're running into UDP throughput limits, one solution would be to increase the the socket option's receive buffer size. (not sure if this is compile-time or runtime config... could we just increase this by default in SC??)

In the meantime, `s.sync` seems like the standard option for throttling server-side message backlog.

Comment (Spacechild1) [reply testsuite/classlibrary/TestCoreUGens.sc:391]: > So, if we're running into UDP throughput limits, one solution would be to increase the the socket option's receive buffer size. 

See https://github.com/supercollider/supercollider/issues/5993 :)

However, I agree with @JordanHendersonMusic that if you want to have reliable communication, TCP is the way to go!

>  So if we can pin down in some degree of detail why TCP allows higher throughput, we could document it as best practices.

With UDP you need to artificially throttle the traffice with `s.sync` or `wait` to prevent socket receive buffer overruns. With TCP this is not necessary, which leads to the curious situation that in some situations you get higher throughput, even though TCP is the slower protocol.
